syntax = "proto3";
package monitor.proto;

import "google/protobuf/timestamp.proto";

// ========== 基础类型 ==========

// 时间范围
message TimeRange {
    google.protobuf.Timestamp start_time = 1;
    google.protobuf.Timestamp end_time = 2;
}

// 分页参数
message Pagination {
    int32 page = 1;      // 页码，从1开始
    int32 page_size = 2; // 每页大小，默认100
}

// 排序方式
enum SortOrder {
    DESC = 0;  // 降序（默认）
    ASC = 1;   // 升序
}

// 服务器状态
enum ServerStatus {
    ONLINE = 0;
    OFFLINE = 1;
}

// ========== 请求消息 ==========

// 时间段性能数据查询请求
message QueryPerformanceRequest {
    string server_name = 1;
    TimeRange time_range = 2;
    Pagination pagination = 3;
}

// 变化率趋势查询请求
message QueryTrendRequest {
    string server_name = 1;
    TimeRange time_range = 2;
    int32 interval_seconds = 3;  // 聚合间隔（秒），0表示不聚合
}

// 异常查询请求
message QueryAnomalyRequest {
    string server_name = 1;       // 可选，空表示查询所有服务器
    TimeRange time_range = 2;
    float cpu_threshold = 3;      // CPU使用率阈值，默认80
    float mem_threshold = 4;      // 内存使用率阈值，默认90
    float disk_threshold = 5;     // 磁盘利用率阈值，默认85
    float change_rate_threshold = 6;  // 变化率阈值，默认0.5（50%）
    Pagination pagination = 7;
}

// 评分排序查询请求
message QueryScoreRankRequest {
    SortOrder order = 1;
    Pagination pagination = 2;
}

// 最新评分查询请求
message QueryLatestScoreRequest {
    // 无参数，返回所有服务器最新数据
}

// 详细数据查询请求
message QueryDetailRequest {
    string server_name = 1;
    TimeRange time_range = 2;
    Pagination pagination = 3;
}

// ========== 数据记录 ==========

// 性能数据记录
message PerformanceRecord {
    string server_name = 1;
    google.protobuf.Timestamp timestamp = 2;
    // CPU指标
    float cpu_percent = 10;
    float usr_percent = 11;
    float system_percent = 12;
    float nice_percent = 13;
    float idle_percent = 14;
    float io_wait_percent = 15;
    float irq_percent = 16;
    float soft_irq_percent = 17;
    // 负载指标
    float load_avg_1 = 20;
    float load_avg_3 = 21;
    float load_avg_15 = 22;
    // 内存指标
    float mem_used_percent = 30;
    float mem_total = 31;
    float mem_free = 32;
    float mem_avail = 33;
    // 磁盘指标
    float disk_util_percent = 40;
    // 网络指标
    float send_rate = 50;
    float rcv_rate = 51;
    // 评分
    float score = 60;
    // 变化率
    float cpu_percent_rate = 70;
    float usr_percent_rate = 71;
    float system_percent_rate = 72;
    float io_wait_percent_rate = 73;
    float load_avg_1_rate = 74;
    float load_avg_3_rate = 75;
    float load_avg_15_rate = 76;
    float mem_used_percent_rate = 77;
    float disk_util_percent_rate = 78;
    float send_rate_rate = 79;
    float rcv_rate_rate = 80;
}

// 异常记录
message AnomalyRecord {
    string server_name = 1;
    google.protobuf.Timestamp timestamp = 2;
    string anomaly_type = 3;      // CPU_HIGH, MEM_HIGH, DISK_HIGH, RATE_SPIKE
    string severity = 4;          // WARNING, CRITICAL
    float value = 5;              // 异常值
    float threshold = 6;          // 阈值
    string metric_name = 7;       // 指标名称
}

// 服务器评分摘要
message ServerScoreSummary {
    string server_name = 1;
    float score = 2;
    google.protobuf.Timestamp last_update = 3;
    ServerStatus status = 4;
    // 关键指标快照
    float cpu_percent = 10;
    float mem_used_percent = 11;
    float disk_util_percent = 12;
    float load_avg_1 = 13;
}

// 集群统计信息
message ClusterStats {
    int32 total_servers = 1;
    int32 online_servers = 2;
    int32 offline_servers = 3;
    float avg_score = 4;
    float max_score = 5;
    float min_score = 6;
    string best_server = 7;
    string worst_server = 8;
}


// ========== 响应消息 ==========

// 时间段性能数据查询响应
message QueryPerformanceResponse {
    repeated PerformanceRecord records = 1;
    int32 total_count = 2;
    int32 page = 3;
    int32 page_size = 4;
}

// 变化率趋势查询响应
message QueryTrendResponse {
    repeated PerformanceRecord records = 1;
    int32 interval_seconds = 2;
}

// 异常查询响应
message QueryAnomalyResponse {
    repeated AnomalyRecord anomalies = 1;
    int32 total_count = 2;
    int32 page = 3;
    int32 page_size = 4;
}

// 评分排序查询响应
message QueryScoreRankResponse {
    repeated ServerScoreSummary servers = 1;
    int32 total_count = 2;
    int32 page = 3;
    int32 page_size = 4;
}

// 最新评分查询响应
message QueryLatestScoreResponse {
    repeated ServerScoreSummary servers = 1;
    ClusterStats cluster_stats = 2;
}

// ========== 详细数据记录 ==========

// 网络详细数据
message NetDetailRecord {
    string server_name = 1;
    string net_name = 2;
    google.protobuf.Timestamp timestamp = 3;
    // 错误和丢弃统计
    uint64 err_in = 10;
    uint64 err_out = 11;
    uint64 drop_in = 12;
    uint64 drop_out = 13;
    // 速率指标
    float rcv_bytes_rate = 20;
    float snd_bytes_rate = 21;
    float rcv_packets_rate = 22;
    float snd_packets_rate = 23;
    // 变化率
    float rcv_bytes_rate_rate = 30;
    float snd_bytes_rate_rate = 31;
    float err_in_rate = 32;
    float err_out_rate = 33;
    float drop_in_rate = 34;
    float drop_out_rate = 35;
}

// 磁盘详细数据
message DiskDetailRecord {
    string server_name = 1;
    string disk_name = 2;
    google.protobuf.Timestamp timestamp = 3;
    // 性能指标
    float read_bytes_per_sec = 10;
    float write_bytes_per_sec = 11;
    float read_iops = 12;
    float write_iops = 13;
    float avg_read_latency_ms = 14;
    float avg_write_latency_ms = 15;
    float util_percent = 16;
    // 变化率
    float read_bytes_per_sec_rate = 20;
    float write_bytes_per_sec_rate = 21;
    float read_iops_rate = 22;
    float write_iops_rate = 23;
    float util_percent_rate = 24;
}

// 内存详细数据
message MemDetailRecord {
    string server_name = 1;
    google.protobuf.Timestamp timestamp = 2;
    // 内存指标 (单位: GB)
    float total = 10;
    float free = 11;
    float avail = 12;
    float buffers = 13;
    float cached = 14;
    float swap_cached = 15;
    float active = 16;
    float inactive = 17;
    float active_anon = 18;
    float inactive_anon = 19;
    float active_file = 20;
    float inactive_file = 21;
    float dirty = 22;
    float writeback = 23;
    float anon_pages = 24;
    float mapped = 25;
    // 变化率
    float total_rate = 30;
    float free_rate = 31;
    float avail_rate = 32;
    float active_rate = 33;
    float inactive_rate = 34;
}

// 软中断详细数据
message SoftIrqDetailRecord {
    string server_name = 1;
    string cpu_name = 2;
    google.protobuf.Timestamp timestamp = 3;
    // 软中断计数
    int64 hi = 10;
    int64 timer = 11;
    int64 net_tx = 12;
    int64 net_rx = 13;
    int64 block = 14;
    int64 irq_poll = 15;
    int64 tasklet = 16;
    int64 sched = 17;
    int64 hrtimer = 18;
    int64 rcu = 19;
    // 变化率
    float hi_rate = 20;
    float timer_rate = 21;
    float net_tx_rate = 22;
    float net_rx_rate = 23;
    float block_rate = 24;
    float sched_rate = 25;
}

// ========== 详细数据查询响应 ==========

message QueryNetDetailResponse {
    repeated NetDetailRecord records = 1;
    int32 total_count = 2;
    int32 page = 3;
    int32 page_size = 4;
}

message QueryDiskDetailResponse {
    repeated DiskDetailRecord records = 1;
    int32 total_count = 2;
    int32 page = 3;
    int32 page_size = 4;
}

message QueryMemDetailResponse {
    repeated MemDetailRecord records = 1;
    int32 total_count = 2;
    int32 page = 3;
    int32 page_size = 4;
}

message QuerySoftIrqDetailResponse {
    repeated SoftIrqDetailRecord records = 1;
    int32 total_count = 2;
    int32 page = 3;
    int32 page_size = 4;
}

// ========== 查询服务定义 ==========

service QueryService {
    // 时间段性能数据查询
    rpc QueryPerformance(QueryPerformanceRequest) returns (QueryPerformanceResponse);
    
    // 变化率趋势查询
    rpc QueryTrend(QueryTrendRequest) returns (QueryTrendResponse);
    
    // 异常数据查询
    rpc QueryAnomaly(QueryAnomalyRequest) returns (QueryAnomalyResponse);
    
    // 评分排序查询
    rpc QueryScoreRank(QueryScoreRankRequest) returns (QueryScoreRankResponse);
    
    // 最新评分查询
    rpc QueryLatestScore(QueryLatestScoreRequest) returns (QueryLatestScoreResponse);
    
    // 详细数据查询
    rpc QueryNetDetail(QueryDetailRequest) returns (QueryNetDetailResponse);
    rpc QueryDiskDetail(QueryDetailRequest) returns (QueryDiskDetailResponse);
    rpc QueryMemDetail(QueryDetailRequest) returns (QueryMemDetailResponse);
    rpc QuerySoftIrqDetail(QueryDetailRequest) returns (QuerySoftIrqDetailResponse);
}
